impl/src/main/java/com/okta/sdk/impl/oauth2/AccessTokenRetrieverServiceImpl.java

```diff
@@ -0,0 +1,184 @@
+import com.okta.commons.http.okhttp.OkHttpRequestExecutorFactory;
+import com.okta.commons.lang.Strings;
+import com.okta.sdk.impl.config.ClientConfiguration;
+import io.jsonwebtoken.Jwts;
+import org.bouncycastle.asn1.pkcs.PrivateKeyInfo;
+import org.bouncycastle.openssl.PEMKeyPair;
+import org.bouncycastle.openssl.PEMParser;
+import org.bouncycastle.openssl.jcajce.JcaPEMKeyConverter;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.*;
+import java.net.HttpURLConnection;
+import java.nio.charset.StandardCharsets;
+import java.security.InvalidKeyException;
+import java.security.KeyPair;
+import java.security.PrivateKey;
+import java.time.Instant;
+import java.time.temporal.ChronoUnit;
+import java.util.Date;
+import java.util.UUID;
+
+public class AccessTokenRetrieverServiceImpl implements AccessTokenRetrieverService {
+    private static final Logger log = LoggerFactory.getLogger(AccessTokenRetrieverServiceImpl.class);
+
+    private static final ObjectMapper objectMapper = new ObjectMapper();
+
+    private ClientConfiguration clientConfiguration;
+
+    public AccessTokenRetrieverServiceImpl(ClientConfiguration clientConfiguration) {
+        this.clientConfiguration = clientConfiguration;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public OAuth2AccessToken getOAuth2AccessToken() throws IOException, InvalidKeyException {
+        log.debug("Getting OAuth2 access token for client id {} from {}",
+            clientConfiguration.getClientId(), clientConfiguration.getBaseUrl());
+
+        String signedJwt = createSignedJWT(clientConfiguration);
+        String scope = String.join(" ", clientConfiguration.getScopes());
+
+        QueryString queryString = new QueryString();
+        queryString.put("grant_type", "client_credentials");
+        queryString.put("client_assertion_type", "urn:ietf:params:oauth:client-assertion-type:jwt-bearer");
+        queryString.put("client_assertion", signedJwt);
+        queryString.put("scope", scope);
+
+        HttpHeaders httpHeaders = new HttpHeaders();
+        httpHeaders.add("Accept", "application/json");
+        httpHeaders.add("Content-Type", "application/x-www-form-urlencoded");
+
+        Request accessTokenRequest = new DefaultRequest(
+            HttpMethod.POST,
+            clientConfiguration.getBaseUrl() + "/oauth2/v1/token",
+            queryString,
+            httpHeaders,
+            new ByteArrayInputStream("".getBytes("UTF-8")),
+            0);
+
+        RequestExecutor requestExecutor = new OkHttpRequestExecutorFactory().create(clientConfiguration);
+
+        Response accessTokenResponse = requestExecutor.executeRequest(accessTokenRequest);
+
+        if (accessTokenResponse != null && accessTokenResponse.getBody() != null) {
+            Reader reader = new InputStreamReader(accessTokenResponse.getBody(), StandardCharsets.UTF_8);
+
+            if (accessTokenResponse.getHttpStatus() == HttpURLConnection.HTTP_OK) {
```

Comment:
you _probably_ want try to use the BaseClient here as this logic _should_ be handled for you already.  I'm not sure how that plays with the oauth related exceptions though

Reply:
BaseClient requires you to implement a bunch of methods and doesnt look straightforward to pull in and use. Furthermore, I would anyway need to examine the response and parse the body into an OAuth2AccessToken object. I find the existing way much more direct and easy.

Reply:
I'll leave this one up to you, the way the SDKs handle the request bodies is to allow for flexibility for the end-user (so they can access arbitrary data, and so updates are not lost between deserializing then reserializing (which happens with current parsing logic)

The scope of this change doesn't require that flexibility as this logic is internal to the SDK.  I mention it mostly because if you could reuse it may save some testing (fewer lines of code to manage, etc).  That said, I'm not 100% sure it would be easier to deal with the different type of error without playing around with it.  But you shouldn't need to implement anything (other than a matching constructor)

You might be able to do something like:
```java
        // reuse instance
        BaseClient tokenClient = new BaseClient(clientConfig, Caches.newDisabledCacheManager()) {};

        try {
            ExtensibleResource responseBody = tokenClient.http()
                .addHeaderParameter("...", "...") // your headers
                .addQueryParameter("...", "...") // your query params
                .setBody(optional) // looks like you don't need a body, but I'm adding it here for completeness
                .post("/token", ExtensibleResource.class);
        } catch (ResourceException e) {

            // this is a bit ugly but it should work, and it something that could be cleaned up 
            DefaultError error = (DefaultError) e.getError();
            String description = error.getString("error_description");
            String errorMessage = error.getString("error");

            // reformat the message to fit the current `Error` object format
            error.setMessage(errorMessage + " - " + description); 
            // ^ also ugly, I don't know what the oauth error messages look like, just using this for an example 
            
            throw new InvalidAuthenticationException(error);
        }
```
Which is very similar to the usage found here: https://github.com/okta/okta-sdk-java#call-other-api-endpoints

It's probably worth taking a look when you get to testing this, there are more or less only two branches here.
Granted I have NOT tested this, just stubbed it out. Your call.

Reply:
Will research on how this could be implemented.